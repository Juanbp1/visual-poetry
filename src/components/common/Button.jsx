import React from 'react';
import PropTypes from 'prop-types';

/**
 * Botón reutilizable con soporte para íconos.
 *
 * @module common/Button
 * @param {Object} props - Propiedades del componente.
 * @param {Function} props.handleClick - Función que se ejecuta al hacer clic en el botón.
 * @param {string} props.ariaLabel - Etiqueta aria para accesibilidad.
 * @param {string} [props.className] - Clase CSS para el botón.
 * @param {string} [props.name] -  Texto que se muestra en el botón.
 * @param {string} [props.text] - Texto del botón.
 * @param {React.ElementType} [props.Icon] -  Un componente React que se renderiza como un ícono.
 * @param {string} [props.iconClassName] - Clase CSS para el ícono.
 * @param {string} [props.title] - Título del botón (tooltip) que aparece al pasar el ratón sobre el botón.
 * @param {string} [props.id] - Un identificador único para el botón.
 *
 * @returns {JSX.Element} Componente de botón.
 *
 * @example
 * // Ejemplo de uso del componente Button
 * import React from 'react';
 * import { FaCoffee } from 'react-icons/fa';
 * import { Button } from '../common'
 *
 * const handleClick = () => {
 *   alert('Botón clicado');
 * }
 *
 * const MyComponent = () => {
 *   return(
 *     <Button
 *      className="btn-primary"
 *      handleClick={handleClick}
 *      title="Botón con ícono"
 *      text="Click me"
 *      Icon={FaCoffee}
 *      iconClassName="icon-class"
 *      ariaLabel="Botón de ejemplo"
 *      id="example-button"
 *      />
 *   };
 * );
 * export default MyComponent;
 */
const Button = ({
  className,
  name,
  text,
  handleClick,
  Icon,
  iconClassName,
  title,
  ariaLabel,
  id,
}) => {
  return (
    <button
      className={className}
      onClick={handleClick}
      title={title}
      name={name}
      aria-label={ariaLabel}
      data-testid={id}
      type="button"
    >
      {Icon && <Icon className={iconClassName} />}
      {text}
    </button>
  );
};
Button.propTypes = {
  handleClick: PropTypes.func.isRequired,
  ariaLabel: PropTypes.string.isRequired,
  className: PropTypes.string,
  iconClassName: PropTypes.string,
  name: PropTypes.string,
  text: PropTypes.string,
  Icon: PropTypes.elementType,
  title: PropTypes.string,
  id: PropTypes.string,
};
export default Button;
