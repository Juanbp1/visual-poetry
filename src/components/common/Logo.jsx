import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * Componente Logo que renderiza un logo SVG con un nombre de texto.
 *
 * @module common/Logo
 * @param {Object} props - Las propiedades del componente.
 * @param {string} props.className - Una clase adicional para aplicar al logo.
 *
 * @returns {JSX.Element} El componente Logo.
 *
 * @example
 * // Ejemplo de uso del componente Logo
 * import React from 'react';
 * import { Logo } from '../common';
 *
 * const MyComponent = () => {
 *   return (<Logo className="primary" />);
 * }
 * export default MyComponent;
 */
const Logo = ({ className = 'header' }) => {
  const logoIconClass = classNames('uiLogo__logoIcon', {
    [`uiLogo__logoIcon--${className}`]: className
  });
  
  const logoNameClass = classNames('uiLogo__logoName', {
    [`uiLogo__logoName--${className}`]: className
  });
  
  return (
    <a
      className="header__link--notUnderlined"
      href="/"
      data-testid={`${className}Logo`}
    >
      <span className="uiLogo__logo" aria-label="logo de la pÃ¡gina web">
        <svg
          className={logoIconClass}
          xmlns="http://www.w3.org/2000/svg"
          viewBox="32798 1839 45 53.598"
          version="1.1"
        >
          <g id="shape">
            <g className="fills">
              <path
                className={logoIconClass}
                d="M32819.572,1839.016C32820.318,1839.007,32821.079,1838.981,32821.822,1839.025C32832.019,1839.625,32840.751,1847.626,32842.625,1857.636C32843.863,1864.248,32842.012,1871.414,32837.387,1876.360C32835.487,1878.392,32834.341,1878.856,32834.343,1881.772L32834.345,1884.584C32834.345,1886.148,32834.263,1887.604,32832.443,1888.215L32832.377,1888.236C32831.957,1888.375,32831.390,1888.400,32830.963,1888.285L32830.906,1888.269C32828.714,1887.668,32828.814,1885.930,32828.813,1884.174L32828.812,1881.680C32828.811,1881.243,32828.853,1880.698,32828.725,1880.278C32827.897,1877.566,32824.267,1877.384,32823.411,1880.062C32823.168,1880.821,32823.273,1881.764,32823.273,1882.558L32823.277,1887.654C32823.276,1888.809,32823.482,1890.056,32822.962,1891.126C32822.363,1892.182,32821.216,1892.749,32820.003,1892.563L32819.947,1892.554C32817.906,1892.236,32817.681,1890.620,32817.680,1888.992L32817.679,1876.255C32817.678,1875.918,32817.712,1875.524,32817.616,1875.199C32816.823,1872.507,32813.213,1872.167,32812.372,1874.895L32812.352,1874.960C32811.917,1876.365,32812.772,1877.977,32811.631,1879.286C32810.101,1881.042,32808.196,1880.243,32806.677,1879.073L32805.299,1877.949C32791.627,1865.971,32798.290,1843.402,32815.924,1839.497C32817.126,1839.231,32818.345,1839.092,32819.572,1839.016ZZM32820.144,1853.839C32818.307,1854.000,32816.672,1854.768,32815.484,1856.200C32811.897,1860.525,32815.243,1867.030,32820.941,1866.720C32826.631,1866.328,32829.122,1859.154,32824.643,1855.367L32823.680,1854.684C32822.718,1854.097,32821.294,1853.672,32820.144,1853.839ZZ"
                fill="black"
              />
            </g>
          </g>
        </svg>
        <p className={logoNameClass}>Visual Poetry</p>
      </span>
    </a>
  );
};
Logo.propTypes = {
  className: PropTypes.string.isRequired,
};
export default Logo;
