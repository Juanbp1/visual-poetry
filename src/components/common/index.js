/**
 * @module common
 * @description Biblioteca de componentes de interfaz de usuario. Proporciona una serie de componentes reutilizables
 * para la construcción de interfaces de usuario dinámicas y responsivas.
 * 
 * @see {@link module:common/Button|Button} - Componente de botón reutilizable con varias propiedades de estilo y eventos.
 * @see {@link module:common/Logo|Logo} - Componente para mostrar el logotipo.
 * @see {@link module:common/ResponsiveComponent|ResponsiveComponent} - Componente que ajusta su renderizado según el tamaño de la pantalla.
 * @see {@link module:common/Textarea|Textarea} - Componente de área de texto con características avanzadas.
 */
export { default as Button } from './Button';
export { default as Logo } from './Logo';
export { default as ResponsiveComponent } from './ResponsiveComponent';
export { default as Textarea } from './Textarea';
