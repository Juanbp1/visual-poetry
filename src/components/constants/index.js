export { default as actions } from './actions';
export { default as actionsFactory } from './actionsFactory';
export { default as actionTypes } from './actionTypes';
// export { default as FAQS_DATA } from './faqsData';
// export { default as SOCIAL_NETWORKS } from './socialNetworks';
// export { default as STEPS_DATA } from './stepsData';
// export { default as uiConstants } from './uiConstants';
