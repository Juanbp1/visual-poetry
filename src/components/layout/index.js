/**
 * @module layout
 * @description Conjunto de componentes principales de la aplicación. Incluye elementos esenciales para la estructura de la interfaz, como el pie de página, la cabecera, la barra lateral y más.
 * 
 * @see {@link module:layout/Footer|Footer} - Componente que representa el pie de página de la aplicación.
 * @see {@link module:layout/header/Header|Header} - Componente que representa el header de la aplicación.
 * @see {@link module:layout/header/DesktopHeader|DesktopHeader} - Componente que representa el header para dispositivos laptop.
 * @see {@link module:layout/header/MobileHeader|MobileHeader} - Componente que representa el header para dispositivos movil.
 * @see {@link module:layout/RenderNav|RenderNav} - Componente que renderiza un menú de navegación.
 * @see {@link module:layout/SideBar|SideBar} - Componente que representa la barra lateral de la aplicación.
 * @see {@link module:layout/SocialNetworks|SocialNetworks} - Componente para mostrar los enlaces a redes sociales.
 * @see {@link module:layout/EditorToolbox|EditorToolbox} - Componente que proporciona herramientas para la edición de texto.
 */

export { default as Footer } from './Footer';
export { default as Header } from './header/Header';
export { default as MobileHeader } from './header/MobileHeader';
export { default as DesktopHeader } from './header/DesktopHeader';
export { default as RenderNav } from './RenderNav';
export { default as SideBar } from './SideBar';
export { default as SocialNetworks } from './SocialNetworks';
export { default as EditorToolbox } from './EditorToolbox';
