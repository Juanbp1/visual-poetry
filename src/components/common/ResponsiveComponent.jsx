import PropTypes from 'prop-types';
import { useState, useEffect, useCallback } from 'react';
import { useMediaQuery } from 'react-responsive';
import { DEVICE_TYPES } from '../constants/uiConstants';

const { TABLET, MOBILE, LAPTOP } = DEVICE_TYPES;
/**
 * Componente que renderiza `componentToRender` basado en el tipo de dispositivo.
 * Utiliza media queries para detectar el tamaño de la pantalla y renderizar el componente asignado,
 * solo si el tipo de dispositivo actual está incluido en `visibleOnDeviceTypes`.
 *
 * @module common/ResponsiveComponent
 * @param {Object} props - Props del componente.
 * @param {Array<string>|string} props.visibleOnDeviceTypes - Tipo(s) de dispositivo para los que el componente debe ser visible.
 * Puede ser un array de tipos de dispositivos o un solo tipo de dispositivo.
 * @param {React.ReactNode} props.componentToRender - El componente que se renderiza si el tipo de dispositivo actual es uno de los `visibleOnDeviceTypes`.
 *
 * @returns {React.ReactElement} El componente renderizado o null si el tipo de dispositivo no está en `visibleOnDeviceTypes`.
 *
 * @example
 * // Ejemplo de uso del componente
 * import React from 'react';
 * import { ResponsiveComponent } from '../common';
 * import { DEVICE_TYPES } from '../constants/uiConstants';
 *
 * const { MOBILE, TABLET } = DEVICE_TYPES;
 *
 * const MyComponent = () => (
 *   <div>
 *     <ResponsiveComponent
 *       visibleOnDeviceTypes={[MOBILE, TABLET]}
 *       componentToRender={<div>Este componente es visible en dispositivos móviles y tabletas.</div>}
 *     />
 *
 *     <ResponsiveComponent
 *       visibleOnDeviceTypes={LAPTOP}
 *       componentToRender={<div>Este componente es visible solo en computadoras portátiles.</div>}
 *     />
 *   </div>
 * );
 *
 * export default MyComponent;
 */
const ResponsiveComponent = ({ visibleOnDeviceTypes, componentToRender }) => {
  const [currentDeviceType, setCurrentDeviceType] = useState(LAPTOP);

  // Media queries para distintos tipos de dispositivos
  const isMobile = useMediaQuery({ query: '(max-width: 430px)' });
  const isTablet = useMediaQuery({
    query: '(min-width: 430px) and (max-width: 1024px)',
  });

  const updateDeviceType = useCallback(() => {
    const currentDeviceType = isMobile ? MOBILE : isTablet ? TABLET : LAPTOP;
    setCurrentDeviceType(currentDeviceType);
  }, [isMobile, isTablet]);

  useEffect(() => {
    updateDeviceType();
    window.addEventListener('resize', updateDeviceType);

    return () => {
      window.removeEventListener('resize', updateDeviceType);
    };
  }, [updateDeviceType]);

  // Devuelve el componente a renderizar si el tipo de dispositivo está en `visibleOnDeviceTypes`
  return visibleOnDeviceTypes.includes(currentDeviceType) && componentToRender;
};
ResponsiveComponent.propTypes = {
  visibleOnDeviceTypes: PropTypes.oneOfType([
    PropTypes.oneOf([MOBILE, TABLET, LAPTOP]),
    PropTypes.arrayOf(PropTypes.oneOf([MOBILE, TABLET, LAPTOP])),
  ]).isRequired,
  componentToRender: PropTypes.node.isRequired,
};

export default ResponsiveComponent;
 