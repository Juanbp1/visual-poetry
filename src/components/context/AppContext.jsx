/**
 * @fileoverview Contexto y Reducer para la aplicación.
 * @module AppContext
 */
import React, { createContext, useContext, useReducer } from 'react';
import actionTypes from '../constants/actionTypes';
import { TEXT_STRINGS } from '../constants/uiConstants';
import actionsFactory from '../constants/actionsFactory';
import defineActions from '../constants/actions';

/**
 * @typedef {Object} AppState
 * @property {Image|null} image - La imagen actual.
 * @property {ImageBorder|null} imageBorder - El borde de la imagen.
 * @property {boolean} imageIsInverted - Indica si la imagen está invertida.
 * @property {Array<ImageGalleryItem>} imageGallery - La galería de imágenes.
 * @property {Object|null} imageSelected - La imagen seleccionada.
 * @property {boolean} imageIsLoaded - Indica si la imagen ha sido cargada.
 * @property {boolean} imageIsProcessed - Indica si la imagen ha sido procesada.
 * @property {boolean} inputIsSetted - Indica si la entrada está configurada.
 * @property {Object} imageBorderPixels - Los píxeles del borde de la imagen.
 * @property {Object} imageData - Datos de la imagen.
 * @property {Object} inputPositionCoordinates - Coordenadas de posición de entrada.
 * @property {string} inputData - Datos de entrada.
 * @property {Object} mainContext - Contexto principal del canvas.
 * @property {Object} secondContext - Segundo contexto del canvas.
 * @property {Object} canvasSize - Tamaño del canvas.
 * @property {Object} styleFont - Estilo de fuente actual.
 * @property {string|null} errorUploadImage - Error al subir la imagen.
 * @property {number} figureThreshold - Umbral para figuras.
 * @property {boolean} loading - Estado de carga.
 * @property {boolean} showGallery - Indica si se muestra la galería.
 * @property {boolean} showGallerySecond - Indica si se muestra la segunda galería.
 * @property {boolean} showGalleryThird - Indica si se muestra la tercera galería.
 */

/**
 * @type {React.Context<{state: AppState, dispatch: React.Dispatch<any>, actions: Object}>}
 */
const AppContext = createContext();

/**
 * Estado inicial del contexto de la aplicación.
 * @type {AppState}
 */
const initialState = {
  image: null,
  imageBorder: null,
  imageIsInverted: false,
  imageGallery: [],
  imageSelected: null,
  imageIsLoaded: false,
  imageIsProcessed: false,
  inputIsSetted: false,
  imageBorderPixels: {},
  imageData: {},
  inputPositionCoordinates: { start: [], end: [] },
  inputData: TEXT_STRINGS.INITIAL_TEXT,
  mainContext: null,
  secondContext: null,
  canvasSize: { canvasWidth: 0, canvasHeight: 0 },
  styleFont: {
    color: '#000000',
    background: '#FF000000',
    font: 'Arial',
    size: 8,
  },
  errorUploadImage: null,
  figureThreshold: 128,
  loading: false,
  showGallery: false,
  showGallerySecond: false,
  showGalleryThird: false,
};

const {
  CHECK_IMAGE_LOADED,
  CHECK_IMAGE_PROCESSED,
  CHECK_INPUT_SETTED,
  OBTAIN_CANVAS_SIZE,
  OBTAIN_IMAGE,
  OBTAIN_IMAGE_BORDER,
  OBTAIN_IMAGE_BORDER_PIXELS,
  OBTAIN_IMAGE_GALLERY,
  OBTAIN_IMAGE_SELECTED,
  OBTAIN_INPUT_POSITION_COORDINATES,
  OBTAIN_MAIN_CONTEXT,
  OBTAIN_SECOND_CONTEXT,
  RESET,
  TOGGLE_INVERT_IMAGE,
  UPDATE_ERROR_UPLOAD_IMAGE,
  UPDATE_FIGURE_THRESHOLD,
  UPDATE_GALLERY_TOGGLE,
  UPDATE_GALLERY_TOGGLE2,
  UPDATE_GALLERY_TOGGLE3,
  UPDATE_IMAGE_GALLERY,
  UPDATE_INPUT_DATA,
  UPDATE_LOADING,
  UPDATE_STYLE_FONT,
} = actionTypes;

/**
 * Función reductora para manejar acciones en el estado de la aplicación.
 * @param {AppState} state - El estado actual de la aplicación.
 * @param {Object} action - La acción que se va a procesar.
 * @param {string} action.type - El tipo de acción que indica qué operación realizar.
 * @param {any} [action.payload] - Los datos asociados a la acción, si los hay.
 * @returns {AppState} - El nuevo estado de la aplicación después de procesar la acción.
 */
const reducer = (state, action) => {
  switch (action.type) {
    case CHECK_IMAGE_LOADED:
      return { ...state, imageIsLoaded: true };

    case CHECK_IMAGE_PROCESSED:
      return { ...state, imageIsProcessed: true };

    case CHECK_INPUT_SETTED:
      return { ...state, inputIsSetted: true };

    case OBTAIN_CANVAS_SIZE:
      return { ...state, canvasSize: action.payload };

    case OBTAIN_IMAGE:
      return { ...state, image: action.payload };

    case OBTAIN_IMAGE_BORDER:
      return { ...state, imageBorder: action.payload };

    case OBTAIN_IMAGE_BORDER_PIXELS:
      if (action.payload && typeof action.payload.getImageData === 'function') {
        const context = action.payload;
        const imageDataPixels = context.getImageData(
          0,
          0,
          context.canvas.width,
          context.canvas.height
        );
        return { ...state, imageBorderPixels: imageDataPixels };
      }
      return { ...state, imageBorderPixels: {} };
    case OBTAIN_IMAGE_GALLERY:
      return { ...state, imageGallery: action.payload };

    case OBTAIN_INPUT_POSITION_COORDINATES:
      return { ...state, inputPositionCoordinates: action.payload };

    case OBTAIN_MAIN_CONTEXT:
      return { ...state, mainContext: action.payload };

    case OBTAIN_SECOND_CONTEXT:
      return { ...state, secondContext: action.payload };

    case OBTAIN_IMAGE_SELECTED:
      return { ...state, imageSelected: action.payload };

    case RESET:
      return { ...initialState };

    case TOGGLE_INVERT_IMAGE:
      return { ...state, imageIsInverted: action.payload };

    case UPDATE_ERROR_UPLOAD_IMAGE:
      return { ...state, errorUploadImage: action.payload };

    case UPDATE_FIGURE_THRESHOLD:
      return { ...state, figureThreshold: action.payload };

    case UPDATE_GALLERY_TOGGLE:
      return { ...state, showGallery: action.payload };

    case UPDATE_GALLERY_TOGGLE2:
      return { ...state, showGallerySecond: action.payload };

    case UPDATE_GALLERY_TOGGLE3:
      return { ...state, showGalleryThird: action.payload };

    case UPDATE_IMAGE_GALLERY:
      const id = action.payload;
      const imageSelected = state.imageGallery[0].map((img) => {
        return { ...img, selected: img.id === id ? !img.selected : false };
      });
      return { ...state, imageGallery: [imageSelected] };

    case UPDATE_INPUT_DATA:
      return { ...state, inputData: action.payload };

    case UPDATE_LOADING:
      return { ...state, loading: !state.loading };

    case UPDATE_STYLE_FONT:
      const updatedStyleFont = {
        ...state.styleFont,
        ...action.payload,
      };
      return {
        ...state,
        styleFont: updatedStyleFont,
      };

    default:
      return state;
  }
};

/**
 * Proveedor del contexto de la aplicación.
 * @param {Object} props - Las propiedades del componente.
 * @param {React.ReactNode} props.children - Los componentes hijos a envolver.
 * @returns {JSX.Element} - El componente Provider con el contexto.
 */
export const AppProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const createAction = actionsFactory(dispatch);

  // Definir las acciones utilizando la fábrica de acciones
  const { imageActions, styleActions, inputActions, othersActions } =
    defineActions(createAction);

  const actions = {
    ...imageActions,
    ...styleActions,
    ...inputActions,
    ...othersActions,
  };

  return (
    <AppContext.Provider value={{ state, dispatch, actions }}>
      {children}
    </AppContext.Provider>
  );
};

/**
 * Hook para utilizar el contexto de la aplicación.
 * @returns {Object} - El contexto de la aplicación que incluye el estado, el despachador y las acciones.
 * @throws {Error} - Lanza un error si el hook se usa fuera del proveedor de contexto.
 */
export const useAppContext = () => {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error(
      'useAppContext debe ser utilizado dentro de un AppProvider'
    );
  }
  return context;
};
