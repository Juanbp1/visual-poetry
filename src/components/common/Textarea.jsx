import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useAppContext } from '../context/AppContext';
import { RiDeleteBin6Line } from 'react-icons/ri';
import { MdContentPaste } from 'react-icons/md';
import { GiInvertedDice5 } from 'react-icons/gi';
import { TEXT_STRINGS, VALIDATION_MESSAGES } from '../constants/uiConstants';
import { Button } from './';
import { useGenerateTextLoremIpsum } from '../../hooks';

/**
 * Componente `Textarea` que permite al usuario editar y manipular texto en un área de texto.
 * Ofrece funcionalidades para borrar, pegar y generar texto lorem ipsum.
 *
 * @module common/Textarea
 * @param {Object} props - Propiedades del componente.
 * @param {string} props.id - Identificador único para el área de texto.
 *
 * @returns {JSX.Element}  El componente de área de texto con botones para manipular el texto.
 *
 * @example
 * // Ejemplo de uso del componente Textarea
 * import React from 'react';
 * import {Textarea} from '../common'
 *
 * const MyComponent = () =>{
 *   return(
 *    <Textarea id="my-textarea" /> *
 *   )
 * }
 * export default MyComponent;
 */
const Textarea = ({ id }) => {
  const { actions, state } = useAppContext();
  const { updateInputDataAction } = actions;
  const { inputData } = state;

  /**
   * Maneja la actualización del texto en el área de texto.
   * @param {React.ChangeEvent<HTMLTextAreaElement>} event - Evento de cambio en el área de texto.
   * @member module:common/Textarea
   */
  const handleUpdateText = (event) => {
    const newText = event.target.value;
    updateInputDataAction(newText);
  };

  /**
   * Maneja la acción de borrar el texto en el área de texto.
   * @member module:common/Textarea
   */
  const handleDeleteText = () => updateInputDataAction('');

  /**
   * Maneja la acción de pegar el texto desde el portapapeles en el área de texto.
   * @member module:common/Textarea
   */
  const handlePasteText = () => {
    if (navigator.clipboard) {
      navigator.clipboard
        .readText()
        .then((clipText) => actions.updateInputDataAction(inputData + clipText))
        .catch(() => {
          alert(TEXT_STRINGS.CLIPBOARD_ERROR_PERMISSION);
        });
    } else {
      alert(TEXT_STRINGS.CLIPBOARD_ERROR_NOT_SUPPORTED);
    }
  };
  const handle = useGenerateTextLoremIpsum();

  /**
   * Maneja la acción de generar texto lorem ipsum y actualizar el área de texto con él.
   * @member module:common/Textarea
   */
  const handleGenerateTextLoremIpsum = () => {
    const newText = handle();
    updateInputDataAction(newText);
  };

  const buttons = [
    {
      handleClick: handleDeleteText,
      title: 'Borrar texto del editor',
      ariaLabel: 'Borrar texto',
      Icon: RiDeleteBin6Line,
    },
    {
      handleClick: handlePasteText,
      title: 'Pegar texto en el Editor',
      ariaLabel: 'Pegar texto',
      Icon: MdContentPaste,
    },
    {
      handleClick: handleGenerateTextLoremIpsum,
      title: 'Generar texto random',
      ariaLabel: 'Generar texto',
      Icon: GiInvertedDice5,
    },
  ];
  useEffect(() => {
    const initialText = TEXT_STRINGS.INITIAL_TEXT;
    updateInputDataAction(initialText);
  }, [updateInputDataAction]);

  return (
    <>
      <div className="textEditor__wrapper" tabIndex="0">
        <div className="textEditor__iconGroup">
          {buttons.map((btn, index) => (
            <div key={index} className="textEditor__iconWrapper">
              <Button
                className="textEditor__icon"
                handleClick={btn.handleClick}
                title={btn.title}
                ariaLabel={btn.ariaLabel}
                Icon={btn.Icon}
              />
            </div>
          ))}
        </div>
        <textarea
          className="textEditor__textarea"
          id={id}
          data-testid={id}
          value={inputData}
          placeholder={VALIDATION_MESSAGES.TEXTAREA_PLACEHOLDER}
          autoComplete="off"
          spellCheck="false"
          onChange={handleUpdateText}
        />
      </div>
    </>
  );
};
Textarea.propTypes = {
  id: PropTypes.string.isRequired,
};

export default Textarea;
