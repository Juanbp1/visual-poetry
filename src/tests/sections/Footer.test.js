import React from 'react';
import {
  cleanup,
  render,
  screen,
  
} from '@testing-library/react';

import { AppProvider } from '../../components/context/AppContext';
import {  Footer } from '../../components/layout';
import { Logo } from '../../components/common';
import { LuGithub } from 'react-icons/lu';
import { SiGmail } from 'react-icons/si';
import { FaLinkedinIn } from 'react-icons/fa';
import { mockUseMediaQuery } from '../mockUseMediaQuery';

import {
  testScrollToTop,
} from '../helpers/userHelpers';

const SOCIAL_NETWORKS = [
  {
    icon: <LuGithub />,
    href: 'https://github.com/your-github-username',
    name: 'github',
  },
  { icon: <SiGmail />, href: 'mailto:youremail@example.com', name: 'gmail' },
  {
    icon: <FaLinkedinIn />,
    href: 'https://www.linkedin.com/in/your-linkedin-profile/',
    name: 'linkedin',
  },
];

describe('El footer debe funcionar correctamente', () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

    test('El logo se muestra correctamente', () => {
      render(
        <AppProvider>
          <Footer>
            <Logo className={'footer'} />
          </Footer>
        </AppProvider>
      );
      const logo = screen.getByTestId('footerLogo');
      expect(logo).toMatchSnapshot();
    });
    test.each(SOCIAL_NETWORKS.map((social) => [social.name, social.href]))(
      'Debería renderizar el link de la red social %s',
      async (name, href) => {
        render(
          <AppProvider>
            <Footer />
          </AppProvider>
        );
   
        const link = screen.getByTestId(`${name}Link`);
        expect(link).toHaveAttribute('href', href);
      }
    );
  
    test.each(SOCIAL_NETWORKS.map((social) => [social.name]))(
      'Se deberia mostrar la red social correctamente %s',
      (name) => {
        render(
          <AppProvider>
            <Footer />
          </AppProvider>
        );
   
        const button = screen.getByTestId(`${name}Link`);
        expect(button).toMatchSnapshot();
      } 
    );
    test('Debería desplazarse a la sección frontpage al hacer clic en el botón "Volver al inicio" (mobile)', () => {
      mockUseMediaQuery({ isMobile: true, isTablet: true });
      testScrollToTop('/', 'footerButton');
    });
    test('Debería desplazarse a la sección frontpage al hacer clic en el botón "Volver al inicio"', () => {
      testScrollToTop('/', 'footerButton');
    }); 
  });