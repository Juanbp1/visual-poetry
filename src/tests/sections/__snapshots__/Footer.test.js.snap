// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`El footer debe funcionar correctamente El logo se muestra correctamente 1`] = `
<a
  class="header__link--notUnderlined"
  data-testid="footerLogo"
  href="/"
>
  <span
    aria-label="logo de la pÃ¡gina web"
    class="uiLogo__logo"
  >
    <svg
      class="uiLogo__logoIcon uiLogo__logoIcon--footer"
      version="1.1"
      viewBox="32798 1839 45 53.598"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        id="shape"
      >
        <g
          class="fills"
        >
          <path
            class="uiLogo__logoIcon uiLogo__logoIcon--footer"
            d="M32819.572,1839.016C32820.318,1839.007,32821.079,1838.981,32821.822,1839.025C32832.019,1839.625,32840.751,1847.626,32842.625,1857.636C32843.863,1864.248,32842.012,1871.414,32837.387,1876.360C32835.487,1878.392,32834.341,1878.856,32834.343,1881.772L32834.345,1884.584C32834.345,1886.148,32834.263,1887.604,32832.443,1888.215L32832.377,1888.236C32831.957,1888.375,32831.390,1888.400,32830.963,1888.285L32830.906,1888.269C32828.714,1887.668,32828.814,1885.930,32828.813,1884.174L32828.812,1881.680C32828.811,1881.243,32828.853,1880.698,32828.725,1880.278C32827.897,1877.566,32824.267,1877.384,32823.411,1880.062C32823.168,1880.821,32823.273,1881.764,32823.273,1882.558L32823.277,1887.654C32823.276,1888.809,32823.482,1890.056,32822.962,1891.126C32822.363,1892.182,32821.216,1892.749,32820.003,1892.563L32819.947,1892.554C32817.906,1892.236,32817.681,1890.620,32817.680,1888.992L32817.679,1876.255C32817.678,1875.918,32817.712,1875.524,32817.616,1875.199C32816.823,1872.507,32813.213,1872.167,32812.372,1874.895L32812.352,1874.960C32811.917,1876.365,32812.772,1877.977,32811.631,1879.286C32810.101,1881.042,32808.196,1880.243,32806.677,1879.073L32805.299,1877.949C32791.627,1865.971,32798.290,1843.402,32815.924,1839.497C32817.126,1839.231,32818.345,1839.092,32819.572,1839.016ZZM32820.144,1853.839C32818.307,1854.000,32816.672,1854.768,32815.484,1856.200C32811.897,1860.525,32815.243,1867.030,32820.941,1866.720C32826.631,1866.328,32829.122,1859.154,32824.643,1855.367L32823.680,1854.684C32822.718,1854.097,32821.294,1853.672,32820.144,1853.839ZZ"
            fill="black"
          />
        </g>
      </g>
    </svg>
    <p
      class="uiLogo__logoName uiLogo__logoName--footer"
    >
      Visual Poetry
    </p>
  </span>
</a>
`;

exports[`El footer debe funcionar correctamente Se deberia mostrar la red social correctamente github 1`] = `
<a
  class="socialMedia__socialIcon socialMedia__socialIcon--footer"
  data-testid="githubLink"
  href="https://github.com/your-github-username"
>
  <svg
    fill="none"
    height="1em"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M15 22v-4a4.8 4.8 0 0 0-1-3.5c3 0 6-2 6-5.5.08-1.25-.27-2.48-1-3.5.28-1.15.28-2.35 0-3.5 0 0-1 0-3 1.5-2.64-.5-5.36-.5-8 0C6 2 5 2 5 2c-.3 1.15-.3 2.35 0 3.5A5.403 5.403 0 0 0 4 9c0 3.5 3 5.5 6 5.5-.39.49-.68 1.05-.85 1.65-.17.6-.22 1.23-.15 1.85v4"
    />
    <path
      d="M9 18c-4.51 2-5-2-7-2"
    />
  </svg>
</a>
`;

exports[`El footer debe funcionar correctamente Se deberia mostrar la red social correctamente gmail 1`] = `
<a
  class="socialMedia__socialIcon socialMedia__socialIcon--footer"
  data-testid="gmailLink"
  href="mailto:youremail@example.com"
>
  <svg
    fill="currentColor"
    height="1em"
    role="img"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title />
    <path
      d="M24 5.457v13.909c0 .904-.732 1.636-1.636 1.636h-3.819V11.73L12 16.64l-6.545-4.91v9.273H1.636A1.636 1.636 0 0 1 0 19.366V5.457c0-2.023 2.309-3.178 3.927-1.964L5.455 4.64 12 9.548l6.545-4.91 1.528-1.145C21.69 2.28 24 3.434 24 5.457z"
    />
  </svg>
</a>
`;

exports[`El footer debe funcionar correctamente Se deberia mostrar la red social correctamente linkedin 1`] = `
<a
  class="socialMedia__socialIcon socialMedia__socialIcon--footer"
  data-testid="linkedinLink"
  href="https://www.linkedin.com/in/your-linkedin-profile/"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M100.28 448H7.4V148.9h92.88zM53.79 108.1C24.09 108.1 0 83.5 0 53.8a53.79 53.79 0 0 1 107.58 0c0 29.7-24.1 54.3-53.79 54.3zM447.9 448h-92.68V302.4c0-34.7-.7-79.2-48.29-79.2-48.29 0-55.69 37.7-55.69 76.7V448h-92.78V148.9h89.08v40.8h1.3c12.4-23.5 42.69-48.3 87.88-48.3 94 0 111.28 61.9 111.28 142.3V448z"
    />
  </svg>
</a>
`;
